test/src/it/scala/RequeueWithExpiryActionIntegrationTest.scala

```diff
@@ -0,0 +1,109 @@
+
+  implicit val ec: ExecutionContext = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(300))
+  implicit val cs: ContextShift[IO] = IO.contextShift(ec)
+  implicit val timer: Timer[IO]     = IO.timer(ec)
+  val requeuePolicy = RequeuePolicy(maximumProcessAttempts = 5, requeueAfter = 2.seconds)
+
+  def withTestFixture[F[_]](onRequeueExpiryAction: String => IO[ConsumeAction])
+                              (test: TestFixture => IO[Unit]): Unit = {
+
+    implicit val payloadMarshaller: PayloadMarshaller[String] = StringPayloadMarshaller
+    implicit val payloadUnmarshaller: PayloadUnmarshaller[String] = StringPayloadUnmarshaller
+
+    val rawConfig = ConfigFactory.load("bucky")
+    val config =
+      AmqpClientConfig(
+        rawConfig.getString("rmq.host"),
+        rawConfig.getInt("rmq.port"),
+        rawConfig.getString("rmq.username"),
+        rawConfig.getString("rmq.password"))
+
+    val exchangeName = ExchangeName(UUID.randomUUID().toString)
+    val routingKey = RoutingKey(UUID.randomUUID().toString)
+    val queueName = QueueName(UUID.randomUUID().toString)
+    val deadletterQueueName = QueueName(s"${queueName.value}.dlq")
+
+    val declarations = List(
+      Exchange(exchangeName).binding(routingKey -> queueName)
+    ) ++ requeue.requeueDeclarations(queueName, routingKey)
+
+    AmqpClient[IO](config).use { client =>
+      val handler = StubHandlers.requeueRequeueHandler[IO, String]
+      val dlqHandler = StubHandlers.ackHandler[IO, String]
+
+      Resource.liftF(client.declare(declarations)).flatMap(_ =>
+        for {
+          _ <-  client.registerRequeueConsumerOf[String](
+            queueName = queueName,
+            handler = handler,
+            requeuePolicy = requeuePolicy,
+            onRequeueExpiryAction = onRequeueExpiryAction
+          )
+          _ <- client.registerConsumerOf(deadletterQueueName, dlqHandler)
+        }
+          yield ()
+      ).use { _ =>
+        val pub = client.publisher()
+        val pcb = publishCommandBuilder[String](implicitly).using(exchangeName).using(routingKey)
+        val fixture = TestFixture(handler, dlqHandler, pcb, pub)
+        test(fixture)
+      }
+    }.unsafeRunSync()
+  }
+
+
+  test("Should DeadLetter after setting onRequeueExpiryAction to DeadLetter on expiry") {
+
+    var wasHit = false
+
+    val expiryAction = (_: String) => IO {
+      wasHit = true
+      DeadLetter
+    }
+
+    withTestFixture(expiryAction) { testFixture =>
+      val publishCommand =
+        testFixture.publishCommandBuilder.toPublishCommand("hello, world!")
+
+      for {
+        _ <- testFixture.publisher(publishCommand)
+      }
+        yield eventually {
+          testFixture.stubHandler.receivedMessages.size should be (requeuePolicy.maximumProcessAttempts)
+          testFixture.dlqHandler.receivedMessages.size shouldBe 1
+          wasHit shouldBe true
```

Comment:
I wonder if it's possible to assert that the action only happens after the final attempt (the requeue expiry). For example use a real handler instead of a stub handler. The handler could add some value (e.g. false) to a list, and the expiry action modifies the last value of the list (flips it to true) so you expect your list to be List(false, false, false, false, true) or something like that.